csharp/client/Dh_NetClient/arrow_util/ArrowUtil.cs

```diff
@@ -0,0 +1,168 @@
+    var dtc = new ArrowDataTypeComparer(lhs);
+    rhs.Accept(dtc);
+    return dtc.Result;
+  }
+
+  public static ArrowTable ToArrowTable(IClientTable clientTable) {
+    var ncols = clientTable.NumCols;
+    var nrows = clientTable.NumRows;
+    var columns = new List<ArrowColumn>();
+
+    for (var i = 0; i != ncols; ++i) {
+      var columnSource = clientTable.GetColumn(i);
+      var arrowArray = ArrowArrayConverter.ColumnSourceToArray(columnSource, nrows);
+      var field = clientTable.Schema.GetFieldByIndex(i);
+      var column = new ArrowColumn(field, [arrowArray]);
+      columns.Add(column);
+    }
+
+    return new ArrowTable(clientTable.Schema, columns);
+  }
+
+  public static string Render(ArrowTable table, bool wantHeaders, bool wantLineNumbers) {
+    var sw = new StringWriter();
+    var numCols = table.ColumnCount;
+
+    var separator = "";
+
+    if (wantHeaders) {
+      var headers = table.Schema.FieldsList.Select(f => f.Name);
+      if (wantLineNumbers) {
+        headers = headers.Prepend("[Row]");
+      }
+
+      sw.Write(string.Join('\t', headers));
+      separator = "\n";
+    }
+
+    var enumerables = Enumerable.Range(0, numCols)
+      .Select(i => MakeScalarEnumerable(table.Column(i).Data).GetEnumerator())
+      .ToArray();
+    var hasMore = new bool[numCols];
+
+    int rowNum = 0;
+
+    var build = new List<object>();
+
+    while (true) {
+      for (var i = 0; i != numCols; ++i) {
+        hasMore[i] = enumerables[i].MoveNext();
+      }
+
+      if (!hasMore.Any(x => x)) {
+        break;
+      }
+
+      build.Clear();
+
+      if (wantLineNumbers) {
+        build.Add($"[{rowNum}]");
+      }
+
+      for (var i = 0; i != numCols; ++i) {
+        if (!hasMore[i]) {
+          build.Add("[exhausted]");
+          continue;
+        }
+        var current = enumerables[i].Current;
+        build.Add(current ?? "[null]");
+      }
+
+      sw.Write(separator);
+      sw.Write(string.Join('\t', build));
+      separator = "\n";
+      ++rowNum;
+    }
+
+    foreach (var e in enumerables) {
+      e.Dispose();
+    }
+    return sw.ToString();
+  }
+
+  public static IEnumerable<object> MakeScalarEnumerable(Apache.Arrow.ChunkedArray chunkedArray) {
+    var numArrays = chunkedArray.ArrayCount;
+    var visitor = new ScalarEnumerableVisitor();
+    for (var i = 0; i != numArrays; ++i) {
+      var array = chunkedArray.ArrowArray(i);
+      array.Accept(visitor);
+      foreach (var result in visitor.Result) {
+        yield return result;
+      }
+    }
+  }
+
+  private class ScalarEnumerableVisitor : Apache.Arrow.IArrowArrayVisitor,
+    IArrowArrayVisitor<ListArray> {
+    public IEnumerable Result = Array.Empty<object>();
+
+    public void Visit(IArrowArray array) {
+      Result = (IEnumerable)array;
+    }
+
+    public void Visit(ListArray array) {
+      Result = ListArrayHelper(array);
+    }
+
+    private IEnumerable ListArrayHelper(ListArray array) {

```

Comment:
WIP, what is ListArray? Is that literally when a column has arrays as content or is it related to barrage?

Reply:
`ListArray` is an Arrow data type, and yes, in this case it has to do when a column has arrays as content, as might happen with a `group_by`. This code is pretty intricate/hairy so I can help explain the details of it if you want to dive into it.